client/my-sites/checklist/wpcom-checklist/index.js

```diff
@@ -332,22 +333,36 @@ function getContactPage( posts ) {
 	);
 }
 
-export default connect(
-	state => {
-		const siteId = getSelectedSiteId( state );
-		const siteSlug = getSiteSlug( state, siteId );
-
+const getTaskUrls = createSelector(
+	( state, siteId ) => {
 		const posts = getPostsForQuery( state, siteId, query );
 
 		const firstPost = find( posts, { type: 'post' } );
+		const siteSlug = getSiteSlug( state, siteId );
 		const contactPage = getContactPage( posts );
 
-		const user = getCurrentUser( state );
-
-		const taskUrls = {
+		return {
 			post_published: compact( [ '/post', siteSlug, get( firstPost, [ 'ID' ] ) ] ).join( '/' ),
 			contact_page_updated: [ '/page', siteSlug, get( contactPage, [ 'ID' ], 2 ) ].join( '/' ),
 		};
+	},
+	(
+		state,
+		siteId,
+		siteSlug = getSiteSlug( state, siteId ),
+		posts = getPostsForQuery( state, siteId, query ),
+		contactPage = getContactPage( posts )
+	) => [ siteId, siteSlug, posts, contactPage ]
```

Comment:
The only dependency we really need here is `posts`. Why? The contract of `createSelector` is:
- I'll keep a map of `args -> return_value` and return a memoized value for every `args` I know. In our case, `args` are `[ siteId ]`.
- every time dependencies change (i.e., the deps array is not shallow-equal to the previous value), it means the memoized values are invalidated. Therefore I'll clear the memoization map and recalculate the results when requested.

This means that `siteId` and `siteSlug` are not good dependencies to have: calling the selectors with a different `siteId` value will forget the previous memoized value.
```
selector( 'site1' );
selector( 'site2' );
selector( 'site1' );
```
won't remember the value for `site1`.

`contactPage` depends on `posts` and nothing else. Therefore it's not an independent dependency. Harmless, but redundant.

That leaves us with `posts`. Theoretically, `siteSlug` could also change for the same `siteId` if we received a new, different `state.sites.items` value. But it practice that will probably never happen -- except edge cases like having just bought and mapped a new primary domain.

Reply:
Thanks, that's an excellent explanation! That will definitely help in the future, whenever I need to write another `createSelector`-based selector.

Reply:
@sgomes in case you haven't come across it yet either there's `treeSelect` which @samouri created to address a sneaky problem particularly on site-dependent selectors: once that cache flushes it's gone even if we flip back and forth between sites and return to the first one.

probably not worth the changeover in this PR, but if it would be helpful to have real memoization instead of a last-used-cache we can think about that one too.

Reply:
Thanks, @dmsnell ! I'll take a look.